### PR TITLE
Gui: take in account module-path argument

### PR DESCRIPTION
**Pull request by ein-shved**
Use paths passed with `--module-path` argument to search for preference packs

I'am [working](https://github.com/NixOS/nixpkgs/pull/347776) on process of introducing the "NIX-way" of configuration of freecad. I'am trying to introduce the `modules` configuration option which allows to pass the downloaded by nix modules to freecad in nix pure way. I'am using the `--module-path` argument for that. It works ok with python modules with `Init.py` fine out the box, but does not work with Preferences Packs, like OpenTheme. 

This patch adds loading Preferences Packs and temlates from paths passed by `--module-path` arguments.

### CODE REVIEW

**Comment by 3x380V** [comment]
@ein-shved, any chance review comment will be taken into account? (expected that before marked ready for review)

**Comment by ein-shved** [comment]
> @ein-shved, any chance review comment will be taken into account? (expected that before marked ready for review)

Sorry, but I cannot see any comments. Don't you forget to submit your comments?

File: src/Gui/PreferencePackManager.cpp
Comment (wwmayer) [inline src/Gui/PreferencePackManager.cpp:239]: What happens with userModPath? Shouldn't this be added to result?

Comment (ein-shved) [reply src/Gui/PreferencePackManager.cpp:239]: Oh! Of course it should! Strange, but this mistake did not brake previous behavior.

Comment (3x380V) [reply src/Gui/PreferencePackManager.cpp:239]: To extend that question: there are about dozen other places in both C++ and Python code constructing path as `fs::path(App::Application::getUserAppDataDir()) / "Mod"`. Are those irrelevant to this PR intend?

**Comment by 3x380V** [comment]
> Sorry, but I cannot see any comments. Don't you forget to submit your comments?

Ahh, I'm sorry! The exactly same [issue described here](https://github.com/orgs/community/discussions/10369) affected this conversation. Github can surprise me almost on montly basis...

**Comment by ein-shved** [comment]
> > Sorry, but I cannot see any comments. Don't you forget to submit your comments?
> 
> Ahh, I'm sorry! The exactly same [issue described here](https://github.com/orgs/community/discussions/10369) affected this conversation. Github can surprise me almost on montly basis...

Eah... they tries to be like gerrit, but doing it in their own awful way... as usual as microsoft doing things.

File: src/Gui/PreferencePackManager.cpp
Comment (ein-shved) [reply src/Gui/PreferencePackManager.cpp:239]: @3x380V  Actually, I ripgrepped all C++ code and except places of this change, there is one place left with "copy-to-clipboard" functionality of about dialog. This copy-to-clipboard functionality uses the folder names under the `"Mod"` as modules names, which is not suitable for  `--module-path` argument, because those names are not standardized. E.g. for nix the folders names are looks like `/nix/store/6n17hn0wj83ck9jg7nd8skaqf453q27d-source`.

Comment (ein-shved) [reply src/Gui/PreferencePackManager.cpp:239]: Done

File: src/Gui/DlgPreferencePackManagementImp.cpp
Comment (ein-shved) [reply src/Gui/DlgPreferencePackManagementImp.cpp:77]: Applied. But auto-formatter did not change this place - it updated code in `[PreferencePackManager.cpp](https://github.com/FreeCAD/FreeCAD/pull/17660/files#diff-ce00f93341f7f2309c1762e58afdf54f7157317f7402040ae86a6aeb191d3354)`

**Review by wwmayer — APPROVED** [review]
I think it's all OK now.

File: src/Gui/PreferencePackManager.cpp
Comment (adrianinsaval) [reply src/Gui/PreferencePackManager.cpp:239]: does this mean that copy to clipboard will not be giving accurate information anymore?

Comment (ein-shved) [reply src/Gui/PreferencePackManager.cpp:239]: It never did it before. You can check: try to run freecad with `--module-path` pointing to the correct module with Init.py outside the standard paths. The module will work, but will not appear in the clipboard. Probably, we should open an issue for that.

File: src/Gui/DlgPreferencePackManagementImp.cpp
Comment (3x380V) [reply src/Gui/DlgPreferencePackManagementImp.cpp:77]: That is strange, here it looks like:
```
+    for (const auto& modDirectory : modDirectories) {
+        if (fs::exists(modDirectory) && fs::is_directory(modDirectory)) {
+            for (const auto& mod : fs::directory_iterator(modDirectory)) {
+                auto packs = getPacksFromDirectory(mod);
+                if (!packs.empty()) {
+                    auto modName = mod.path().filename().string();
+                    installedPacks.emplace(modName, packs);
+                }
```
Well, nevermind...

**Review by 3x380V — APPROVED** [review]

**Comment by 3x380V** [comment]
@ein-shved, for the sake of completeness, here's what clang-format does with your patch: https://github.com/FreeCAD/FreeCAD/commit/d2eb664f993e0470d6f313096f1bf6126026f201

**Comment by ein-shved** [comment]
> @ein-shved, for the sake of completeness, here's what clang-format does with your patch: [d2eb664](https://github.com/FreeCAD/FreeCAD/commit/d2eb664f993e0470d6f313096f1bf6126026f201)

Hmmm I guess you right. Maybe I missed something - it is not very convenient to apply format only to changed lines. Will recheck later.

**Comment by 3x380V** [comment]
That was mostly manual work. The whole point is to have as straightforward history as possible. Without this change eventual `git-blame` would show also commit which changes formatting after #16077 gets merged, while with this change code this commit modified stay untouched.

**Comment by ein-shved** [comment]
> `git-blame` would show also commit which changes formatting

Yeah that a big problem of applying auto-formatter to legacy projects. But still you are able to configure CI with [clang-format-diff](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/clang-format-diff.py) right now without changing everything.

File: src/Gui/DlgPreferencePackManagementImp.cpp
Comment (ein-shved) [reply src/Gui/DlgPreferencePackManagementImp.cpp]: With [git-clang-format](https://github.com/llvm/llvm-project/blob/main/clang/tools/clang-format/git-clang-format) things become much easier!

```
nix shell nixpkgs#libclang.python nixpkgs#clang-tools
git clang-format HEAD^
git commit -a --amend --no-edit
git push --force
```

File: src/Gui/PreferencePackManager.cpp
Comment (ein-shved) [reply src/Gui/PreferencePackManager.cpp:239]: @adrianinsaval, I added an issue #17843

**Comment by ein-shved** [comment]
@adrianinsaval Hi! Is there anything else needed on my part to apply this change?

**Comment by ein-shved** [comment]
@wwmayer @3x380V @adrianinsaval, Hi! I'm sorry for interrupting, but what have I do to merge this PR?

**Comment by luzpaz** [comment]
We also need documentation for this feature.

**Comment by ein-shved** [comment]
> We also need documentation for this feature.

I think, this is not a feature. This is simply partial fix of invalid behavior.

**Comment by ein-shved** [comment]
@luzpaz, actually, [current documentation](https://wiki.freecad.org/Preference_Packs#Distributing_a_pack) is totally wrong. It says:

> To install a pack manually, use git to clone the package repository into your FreeCAD data directory (enter App.getUserAppDataDir() in the [Python console](https://wiki.freecad.org/Python_console) to get this path), in a subdirectory called "Preference Packs".

But Preference Packs are loaded from subdirectory called "Mod"

**Comment by ein-shved** [comment]
@luzpaz @3x380V @wwmayer @adrianinsaval Please, can you make a decision about this fix.

**Comment by 3x380V** [comment]
@ein-shved, I'm just a mere contributor without possibility to make any decision :) It looks good to me and fixing documentation is a separate PR material. This fix just revealed documentation needs some polishing and it should not block this PR.

**Comment by wwmayer** [comment]
I said it several weeks ago that this PR is fine to merge. But in the meantime an import fix for unicode handling has been merged that causes some merge conflicts.

@ein-shved 
Will it be easy for you to resolve the conflicts?

**Comment by wwmayer** [comment]
Thanks!

**Pull request merged**
